### DISCUSSION TITLE
Basic control of lights and pump

### DISCUSSION THREAD

**Discussion by yunuskink** [iot-root/gardyn-of-eden#79] [General]
I saw that the pump should have a hard limit of 30% PWM. Is that at the raspberry pi zero level or is that electronically encoded. In short, if I supply a constant 5 Volts to pin 18 controlling the pump, will that break something and should I be careful to only ever do PWM of 30%?  

Background: My gardyn sporadically just won't connect to wifi which makes it almost entirely useless. I found when I took it apart that if I jumper pin 2 (5V) and pin 12 (LED) and pin 18 (pump), then those things will switch on. I want to just send wires to those pins outside the gardyn so I can just simply control these things externally with switches or another arduino or pi and skip coding the built-in pi zero at all. I just want to get this thing back together and basically functional and figure out some smart control of these two components later. I don't care about the water level, temp, humidity, or camera stuff. I never used those when they worked anyway. Thanks!

**Comment by iot-root** [comment]
So, the 30% is my speculation as there is an on board PCB temp sensor I am assuming that was intended to stop the pump or maybe lights as a fail safe. I think you would be fine controlling directly if you have the right voltage for the pump. I would still try to limit the on time though. If you need help getting you pi zero going again, jump on discord https://discord.gg/H35VwazSDb and I can help you out.

Likely a few common scenarios: 
1. there was a batch of counterfit sdcards causing corruption issues,
2. an bad update in march of 2024 broke a bunch of code , you might be able to recover by connecting a micro HMDI and keyboard then check that the gardyn services are running with no errors. If there are errors related to missing arrow or cors_flask, try running `sudo pip3 install arrow cors_flask` then rebooting.

**Comment by ragrossman** [comment]
Hey yunuskink, wondering if you managed to set up external switches for the system. If so, would you be open to telling me how you did it? Having the same wifi issue and would rather skip coding if possible.

**Reply by yunuskink** [reply]
Yes, I got it to work with an old arduino and real-time clock module I stole from my kid's arduino kit I bought him a while back. You could do this with any microcontroller though, including the pi zero. All it does is follows a schedule and sends a PWM signal to pins 12 and 18 on the board for the gardyn. I just unscrewed and pulled out the pi zero totally. I had AI write the code and it is below. The schedule is tuned to when I want bright lights in the room, I don't know if it is any good for the plants at all. I want to add some features like a simple manual switch for the lights and also am considering adding a sensor that detected the sound intensity of the room and would blink the gardyn lights when my kids are being too loud!

In retrospect, I suspect the rootuser is right and that a quick software fix of the SD card might be the problem since the pi seemed fine, but at this point I just wanted my plants to get some light and water and didn't really care much about the cameras or the app anymore. My water level sensor had been destroyed by this time and my temperature and humidity sensor also wasn't looking so good.

I just used the power that was supplied to the pi to run the arduino. Happy to share more details if you want to go this route!

```
#include <Wire.h>
#include <RTClib.h>

RTC_DS1307 rtc;

const int pwmPin_pump = 9;
const int pwmPin_lights = 10;

// Define schedule times and corresponding PWM values
struct Schedule_lights {
  int hour;
  int minute;
  int pwmValue;
};

struct Schedule_pump {
  int hour;
  int minute;
  int pwmValue;
  int duration;
};

Schedule_lights schedule_lights_weekend[] = {
  {6, 0, 0},  // 6:00 AM, 0 minutes, PWM = 0 (lights off, 255 is maximum)
  {7, 0, 24}, 
  {8, 0, 128},
  {9, 0, 128},  
  {10, 0, 128},  
  {13, 0, 128},  
  {14, 0, 128},  
  {17, 0, 255},  
  {19, 0, 128},  
  {20, 0, 128},  
  {21, 0, 0},  
};

Schedule_lights schedule_lights_weekday[] = {
  {6, 45, 128}, 
  {7, 0, 255}, 
  {8, 0, 128},
  {9, 0, 128},  
  {10, 0, 128},  
  {13, 0, 128},  
  {14, 0, 128},  
  {17, 0, 255},  
  {19, 0, 128},  
  {20, 0, 128},  
  {21, 0, 0},  
};

Schedule_pump schedule_pump[] = {
  {8, 0, 75, 15},   // 8:00 AM, PWM = 75, duration = 15 minutes
  {12, 0, 75, 15},  // 12:00 PM, PWM = 75, duration = 15 minutes
  {18, 0, 75, 15},  // 6:00 PM, PWM = 75, duration = 15 minutes
  {22, 0, 75, 15}   // 10:00 PM, PWM = 75, duration = 15 minutes
};

void setup() {
  Serial.begin(9600);
  if (!rtc.begin()) {
    Serial.println("Couldn't find RTC");
    while (1);
  }


  Serial.println("RTC is NOT running, let's set the time!");
  // Set the RTC to the date & time this sketch was compiled
  rtc.adjust(DateTime(F(__DATE__), F(__TIME__)));

  pinMode(pwmPin_pump, OUTPUT);
  pinMode(pwmPin_lights, OUTPUT);
}

void loop() {
  

  DateTime now = rtc.now();
  int dayOfWeek = now.dayOfTheWeek(); // 0 = Sunday, 1 = Monday, ..., 6 = Saturday

  Schedule_lights* schedule_lights;
  int schedule_lights_size;

  if (dayOfWeek == 0 || dayOfWeek == 6) { // Weekend
    schedule_lights = schedule_lights_weekend;
    schedule_lights_size = sizeof(schedule_lights_weekend) / sizeof(schedule_lights_weekend[0]);
  } else { // Weekday
    schedule_lights = schedule_lights_weekday;
    schedule_lights_size = sizeof(schedule_lights_weekday) / sizeof(schedule_lights_weekday[0]);
  }

  // Print the current time and day of the week
  Serial.print("Current time: ");
  Serial.print(now.hour());
  Serial.print(":");
  Serial.print(now.minute());
  Serial.print(":");
  Serial.println(now.second());

  Serial.print("Day of the week: ");
  Serial.println(dayOfWeek);

  // Loop for lights to find the most recent time
  int mostRecentIndex = -1;
  for (int i = 0; i < schedule_lights_size; i++) {
    int scheduleMinute = schedule_lights[i].hour * 60 + schedule_lights[i].minute;
    int currentMinute = now.hour() * 60 + now.minute();
    if (currentMinute >= scheduleMinute) {
      mostRecentIndex = i;
    } else {
      break;
    }
  }

  if (mostRecentIndex != -1) {
    analogWrite(pwmPin_lights, schedule_lights[mostRecentIndex].pwmValue);
    Serial.print("Lights PWM set to: ");
    Serial.println(schedule_lights[mostRecentIndex].pwmValue);
  }

  // Loop for pump
  for (int i = 0; i < sizeof(schedule_pump) / sizeof(schedule_pump[0]); i++) {
    int startMinute = schedule_pump[i].hour * 60 + schedule_pump[i].minute;
    int currentMinute = now.hour() * 60 + now.minute();
    int endMinute = startMinute + schedule_pump[i].duration;

    if (currentMinute >= startMinute && currentMinute < endMinute) {
      analogWrite(pwmPin_pump, schedule_pump[i].pwmValue);
      Serial.print("Pump PWM set to: ");
      Serial.println(schedule_pump[i].pwmValue);
      break;
    } else {
      analogWrite(pwmPin_pump, 0);
      Serial.println("Pump PWM set to: 0");
    }
  }

  delay(60000); // Check every minute
}

```
